# Quant Analyzer RAG


---
Refinements:
- Adjusted instructions to make it clear what analysis should be done given the subtasks
- Had to adjust planner agent instructions to produce more relevant subtasks that could be completed with the given data after running the analyzer and seeing what the analysis and results looked like


In [ ]:
# ========================================
# Install Packages
# ========================================
!pip install -q llama-index
!pip install -q llama-index-llms-openai
!pip install -q llama-index-embeddings-openai
!pip install -q openai
!pip install -q pandas numpy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 102.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 303.3/303.3 kB 24.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.0/92.0 kB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.9/63.9 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 329.5/329.5 kB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 53.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.7/150.7 kB 14.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16,

In [ ]:
# ========================================
# Import Libraries
# ========================================
import os
import json
from pathlib import Path
from datetime import datetime
from typing import Dict, Any, List
from getpass import getpass
import glob

# LlamaIndex imports
from llama_index.core import VectorStoreIndex, SimpleDirectoryReader, Document, Settings
from llama_index.embeddings.openai import OpenAIEmbedding
from llama_index.llms.openai import OpenAI as LlamaOpenAI
from llama_index.core.node_parser import SimpleNodeParser
from llama_index.core.prompts import PromptTemplate

In [ ]:
# ========================================
# Set up OpenAI API key
# ========================================
os.environ["OPENAI_API_KEY"] = "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"

In [ ]:
# ========================================
# Upload Financial Statement Files
# ========================================
from google.colab import files

print("📁 Upload your financial statement files (PDFs, Excel, CSV, TXT)")
print("Examples: income_statements.pdf, balance_sheets.xlsx, cash_flow.pdf")
print("=" * 60)
print("TIP: Name files clearly, e.g., 'abc_income_2022.pdf', 'xyz_balance_2022.pdf'")
print("=" * 60)

uploaded = files.upload()

# Create directory for uploads
os.makedirs('financial_statements', exist_ok=True)

# Save uploaded files
for filename in uploaded.keys():
    filepath = f'financial_statements/{filename}'
    with open(filepath, 'wb') as f:
        f.write(uploaded[filename])
    print(f"✅ Saved: {filename} ({len(uploaded[filename])} bytes)")

# Get list of all uploaded files
financial_file_paths = glob.glob('financial_statements/*')
print(f"\n📊 Total financial files uploaded: {len(financial_file_paths)}")

📁 Upload your financial statement files (PDFs, Excel, CSV, TXT)
Examples: income_statements.pdf, balance_sheets.xlsx, cash_flow.pdf
TIP: Name files clearly, e.g., 'abc_income_2022.pdf', 'xyz_balance_2022.pdf'


Saving XYZ_Quarterly_Financials.xlsx to XYZ_Quarterly_Financials (1).xlsx
Saving Company_Annual_Financials_ABC.xlsx to Company_Annual_Financials_ABC (1).xlsx
Saving Company_Balance_Sheet_Annual_ABC.xlsx to Company_Balance_Sheet_Annual_ABC (1).xlsx
Saving Company_Balance_Sheet_Quarterly_ABC.xlsx to Company_Balance_Sheet_Quarterly_ABC (1).xlsx
Saving Company_Quarterly_Financials_ABC.xlsx to Company_Quarterly_Financials_ABC (1).xlsx
Saving XYZ_Annual_Financials.xlsx to XYZ_Annual_Financials (1).xlsx
Saving XYZ_Balance_Sheet_Detailed_Annual.xlsx to XYZ_Balance_Sheet_Detailed_Annual (1).xlsx
Saving XYZ_Balance_Sheet_Quarterly_Detailed.xlsx to XYZ_Balance_Sheet_Quarterly_Detailed (1).xlsx
✅ Saved: XYZ_Quarterly_Financials (1).xlsx (10915 bytes)
✅ Saved: Company_Annual_Financials_ABC (1).xlsx (11418 bytes)
✅ Saved: Company_Balance_Sheet_Annual_ABC (1).xlsx (11482 bytes)
✅ Saved: Company_Balance_Sheet_Quarterly_ABC (1).xlsx (11479 bytes)
✅ Saved: Company_Quarterly_Financials_ABC (1).xlsx (1118

In [ ]:
# ========================================
# Configure Company Mapping
# ========================================
print("\n🏢 CONFIGURE COMPANY MAPPING")
print("=" * 60)
print("Map your files to companies (ABC or XYZ)")
print("=" * 60)

# Automatic mapping based on filename
company_mapping = {}

for filepath in financial_file_paths:
    filename = Path(filepath).name.lower()

    # Try to auto-detect company
    if 'abc' in filename:
        company_mapping[filepath] = "ABC"
        print(f"✅ {Path(filepath).name} → ABC")
    elif 'xyz' in filename:
        company_mapping[filepath] = "XYZ"
        print(f"✅ {Path(filepath).name} → XYZ")
    else:
        print(f"⚠️ {Path(filepath).name} → Could not auto-detect company")
        company_mapping[filepath] = "Unknown"

print(f"\n📋 Company mapping complete!")
print(f"   ABC files: {len([f for f in company_mapping.values() if f == 'ABC'])}")
print(f"   XYZ files: {len([f for f in company_mapping.values() if f == 'XYZ'])}")


🏢 CONFIGURE COMPANY MAPPING
Map your files to companies (ABC or XYZ)
✅ Company_Annual_Financials_ABC (1).xlsx → ABC
✅ XYZ_Quarterly_Financials (1).xlsx → XYZ
✅ XYZ_Balance_Sheet_Detailed_Annual (1).xlsx → XYZ
✅ XYZ_Annual_Financials (1).xlsx → XYZ
✅ Company_Balance_Sheet_Quarterly_ABC (1).xlsx → ABC
✅ XYZ_Balance_Sheet_Quarterly_Detailed (1).xlsx → XYZ
✅ Company_Balance_Sheet_Annual_ABC (1).xlsx → ABC
✅ Company_Quarterly_Financials_ABC (1).xlsx → ABC

📋 Company mapping complete!
   ABC files: 4
   XYZ files: 4


In [ ]:
# --------------------
# Configuration
# --------------------
BASE_PATH = "/content"
PLANNER_JSON_PATH = f"{BASE_PATH}/planner_output (1).json"

In [ ]:
# =====================================
# Quant Analyzer Agent Class
# =====================================
class QuantAnalyzerAgent:
    """RAG-based Quantitative Analyzer for financial analysis."""

    def __init__(self, openai_api_key: str = None):
        self.api_key = openai_api_key or os.getenv("OPENAI_API_KEY")

        # Configure LlamaIndex
        Settings.llm = LlamaOpenAI(
            model="gpt-3.5-turbo",
            temperature=0.1,
            api_key=self.api_key
        )
        Settings.embed_model = OpenAIEmbedding(
            model="text-embedding-ada-002",
            api_key=self.api_key
        )
        Settings.node_parser = SimpleNodeParser.from_defaults(
            chunk_size=512,
            chunk_overlap=50
        )

        self.index = None
        self.query_engine = None

    # -----------------------------
    # Document Loading
    # -----------------------------
    def load_financial_documents(self, file_paths: List[str], company_mapping: Dict[str, str]) -> List[Document]:
        """Load financial documents and attach metadata."""
        print("\n🔄 Loading financial documents...")
        documents = []

        for file_path in file_paths:
            path = Path(file_path)
            if not path.exists():
                print(f"⚠️ File not found: {file_path}")
                continue

            try:
                # Infer document type
                fname = path.name.lower()
                if 'income' in fname or 'profit' in fname:
                    doc_type = "income_statement"
                elif 'balance' in fname:
                    doc_type = "balance_sheet"
                elif 'cash' in fname:
                    doc_type = "cash_flow"
                else:
                    doc_type = "financial_statement"

                # Load document
                reader = SimpleDirectoryReader(input_files=[str(path)])
                docs = reader.load_data()

                # Add metadata
                company = company_mapping.get(file_path, "Unknown")
                for doc in docs:
                    doc.metadata.update({
                        "company": company,
                        "document_type": doc_type,
                        "source_file": path.name
                    })

                documents.extend(docs)
                print(f"✅ {path.name} → {company} ({doc_type})")

            except Exception as e:
                print(f"❌ Error loading {file_path}: {e}")

        print(f"\n✅ Loaded {len(documents)} document pages")
        return documents

    # -----------------------------
    # Build Financial Knowledge Base
    # -----------------------------
    def build_financial_index(self, file_paths: List[str], company_mapping: Dict[str, str]):
        """Build vector index and query engine for financial analysis."""
        print("\n📊 Building Financial Knowledge Base...")
        documents = self.load_financial_documents(file_paths, company_mapping)

        if not documents:
            raise ValueError("No documents loaded!")

        print("\n🔄 Creating vector index...")
        self.index = VectorStoreIndex.from_documents(documents, show_progress=True)

        print("🔄 Setting up query engine...")
        self._create_query_engine()
        print("✅ Financial knowledge base ready!")

    def _create_query_engine(self):
        """Create query engine with financial prompt."""
        qa_template = PromptTemplate(
            """You are a Financial Analyst calculating metrics from financial statements.

              Context from financial statements:
              {context_str}

              Question: {query_str}

              Instructions:
              1. Extract exact numbers from the statements
              2. Show all calculations step-by-step
              3. Identify trends across years
              4. Be precise with units (millions, thousands, etc.)
              5. If data is missing, state what's needed

              Your analysis:"""
        )
        self.query_engine = self.index.as_query_engine(
            similarity_top_k=8,
            text_qa_template=qa_template
        )

    # -----------------------------
    # Financial Subtask Analysis
    # -----------------------------
    def analyze_financial_subtask(self, subtask: Dict[str, Any]) -> Dict[str, Any]:
        """Analyze a financial subtask safely."""
        print(f"\n{'='*60}")
        print(f"🔢 {subtask.get('id')}: {subtask.get('title')}")
        print(f"{'='*60}")

        # Build query
        query_parts = [f"Objective: {subtask.get('objective', '')}"]
        normalized_metrics = []

        for metric in subtask.get('metrics', []):
            if isinstance(metric, dict):
                name = metric.get("name", "Unknown Metric")
                formula = metric.get("formula", "N/A")
            else:
                name = str(metric)
                formula = "N/A"
            normalized_metrics.append(name)
            query_parts.append(f"  - {name}: {formula}")

        checklist = subtask.get('checklist', [])
        if checklist:
            query_parts.append("\nSteps:\n" + "\n".join([f"  - {i}" for i in checklist]))

        query = "\n".join(query_parts)
        response = self.query_engine.query(query)

        # Collect sources
        sources = []
        if hasattr(response, 'source_nodes'):
            sources = [
                {
                    "company": node.node.metadata.get('company'),
                    "doc_type": node.node.metadata.get('document_type'),
                    "file": node.node.metadata.get('source_file')
                }
                for node in response.source_nodes[:3]
            ]

        return {
            "subtask_id": subtask.get('id'),
            "subtask_title": subtask.get('title'),
            "metrics": normalized_metrics,
            "analysis": str(response),
            "sources": sources,
            "confidence": "high" if len(sources) >= 3 else "medium"
        }

    def analyze_financial_batch(self, subtasks: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Analyze multiple financial subtasks."""
        print(f"\n🚀 Analyzing {len(subtasks)} financial subtasks...")
        results = []
        for i, st in enumerate(subtasks, 1):
            print(f"\n[{i}/{len(subtasks)}]")
            try:
                results.append(self.analyze_financial_subtask(st))
                print(f"✅ Complete | Confidence: {results[-1]['confidence']}")
            except Exception as e:
                print(f"❌ Error: {e}")
                results.append({"subtask_id": st.get('id'), "error": str(e), "status": "failed"})
        return results

    # -----------------------------
    # Comprehensive Comparison
    # -----------------------------
    def comprehensive_comparison(self, years: List[int] = None) -> Dict[str, Any]:
        """Compare ABC and XYZ financially."""
        print("\n💰 Generating comprehensive financial comparison...")
        years_str = f"for {', '.join(map(str, years))}" if years else "for all years"
        query = f"""
                Compare ABC and XYZ financially {years_str}:

                1. PROFITABILITY: Gross, Net, Operating Margin
                2. EFFICIENCY: Inventory Turnover, Asset Turnover
                3. CASH MANAGEMENT: Cash Conversion Cycle
                4. GROWTH: Revenue & Net Income growth
                5. COSTS: SG&A %, COGS %

                Show all numbers and calculations.
                """
        response = self.query_engine.query(query)
        return {"analysis_type": "comprehensive_comparison", "analysis": str(response)}


print("✅ Quant Analyzer class loaded!")

✅ Quant Analyzer class loaded!


In [ ]:
# ========================================
# Run Complete Financial Analysis
# ========================================
def run_financial_analysis(planner_json_path: str):
    """Run complete quantitative financial analysis using uploaded files."""

    import json
    from datetime import datetime

    print("\n" + "="*80)
    print("🔢 STARTING QUANTITATIVE FINANCIAL ANALYSIS")
    print("="*80)

    # STEP 1: Load Planner Output
    print(f"📂 Loading planner output from: {planner_json_path}")
    with open(planner_json_path, "r") as f:
        planner_output = json.load(f)

    financial_subtasks = planner_output.get('subtasks', {}).get('financial_subtasks', [])

    if not financial_subtasks:
        print("⚠️ Warning: No financial subtasks found in planner output!")
        return None, None

    print(f"\n✅ Loaded {len(financial_subtasks)} financial subtasks:")
    for st in financial_subtasks:
        print(f"   • {st['id']}: {st['title']}")

    # STEP 2: Build Financial Index
    print("\n📊 STEP 2: BUILDING FINANCIAL KNOWLEDGE BASE")
    print("-" * 60)

    analyzer = QuantAnalyzerAgent()
    analyzer.build_financial_index(financial_file_paths, company_mapping)

    # STEP 3: Analyze Subtasks
    print("\n🔍 STEP 3: ANALYZING FINANCIAL METRICS")
    print("-" * 60)

    results = analyzer.analyze_financial_batch(financial_subtasks)

    # STEP 4: Comprehensive Comparison
    print("\n📊 STEP 4: COMPREHENSIVE COMPARISON")
    print("-" * 60)

    comparison = analyzer.comprehensive_comparison(years=[2020, 2021, 2022])

    # STEP 5: Display Results
    print("\n" + "="*80)
    print("📊 FINANCIAL ANALYSIS RESULTS")
    print("="*80)

    for result in results:
        if "error" not in result:
            print(f"\n{'='*60}")
            print(f"📌 {result['subtask_id']}: {result['subtask_title']}")
            print(f"{'='*60}")
            print(f"📊 Metrics: {', '.join(result['metrics'])}")
            print(f"\n{result['analysis']}")
            print(f"\n📚 Confidence: {result['confidence']}")

    # STEP 6: Save results
    output = {
        "planner_output": planner_output,
        "financial_results": results,
        "comparison": comparison,
        "timestamp": datetime.now().isoformat()
    }

    output_file = "financial_analysis_results.json"
    with open(output_file, 'w') as f:
        json.dump(output, f, indent=2)

    print(f"\n💾 Results saved to: {output_file}")

    # Optional: Download in Colab
    from google.colab import files as colab_files
    colab_files.download(output_file)
    print(f"📥 Downloaded: {output_file}")

    return results, comparison
results, comparison = run_financial_analysis(PLANNER_JSON_PATH)


🔢 STARTING QUANTITATIVE FINANCIAL ANALYSIS
📂 Loading planner output from: /content/planner_output (1).json

✅ Loaded 10 financial subtasks:
   • 1: Evaluate Inventory Turnover
   • 2: Analyze Cash Conversion Cycle
   • 3: Assess Gross Margin
   • 4: Evaluate Net Profit Margin
   • 5: Analyze SG&A %
   • 6: Assess ROI
   • 7: Analyze Operating Margin
   • 8: Evaluate ROA
   • 9: Analyze ROE
   • 10: Evaluate Revenue per Employee

📊 STEP 2: BUILDING FINANCIAL KNOWLEDGE BASE
------------------------------------------------------------

📊 Building Financial Knowledge Base...

🔄 Loading financial documents...
✅ Company_Annual_Financials_ABC (1).xlsx → ABC (financial_statement)
✅ XYZ_Quarterly_Financials (1).xlsx → XYZ (financial_statement)
✅ XYZ_Balance_Sheet_Detailed_Annual (1).xlsx → XYZ (balance_sheet)
✅ XYZ_Annual_Financials (1).xlsx → XYZ (financial_statement)
✅ Company_Balance_Sheet_Quarterly_ABC (1).xlsx → ABC (balance_sheet)
✅ XYZ_Balance_Sheet_Quarterly_Detailed (1).xlsx → XYZ (ba

Parsing nodes:   0%|          | 0/8 [00:00<?, ?it/s]

Generating embeddings:   0%|          | 0/77 [00:00<?, ?it/s]

🔄 Setting up query engine...
✅ Financial knowledge base ready!

🔍 STEP 3: ANALYZING FINANCIAL METRICS
------------------------------------------------------------

🚀 Analyzing 10 financial subtasks...

[1/10]

🔢 1: Evaluate Inventory Turnover
✅ Complete | Confidence: high

[2/10]

🔢 2: Analyze Cash Conversion Cycle
✅ Complete | Confidence: high

[3/10]

🔢 3: Assess Gross Margin
✅ Complete | Confidence: high

[4/10]

🔢 4: Evaluate Net Profit Margin
✅ Complete | Confidence: high

[5/10]

🔢 5: Analyze SG&A %
✅ Complete | Confidence: high

[6/10]

🔢 6: Assess ROI
✅ Complete | Confidence: high

[7/10]

🔢 7: Analyze Operating Margin
✅ Complete | Confidence: high

[8/10]

🔢 8: Evaluate ROA
✅ Complete | Confidence: high

[9/10]

🔢 9: Analyze ROE
✅ Complete | Confidence: high

[10/10]

🔢 10: Evaluate Revenue per Employee
✅ Complete | Confidence: high

📊 STEP 4: COMPREHENSIVE COMPARISON
------------------------------------------------------------

💰 Generating comprehensive financial comparison.

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

📥 Downloaded: financial_analysis_results.json


In [ ]:
# =====================================
# Display Summary Statistics
# =====================================
print("\n" + "="*80)
print("✨ FINANCIAL ANALYSIS SUMMARY")
print("="*80)

successful = len([r for r in results if "error" not in r])
failed = len(results) - successful

print(f"\n📊 Total financial subtasks: {len(results)}")
print(f"✅ Successful: {successful}")
print(f"❌ Failed: {failed}")

all_metrics = []
for r in results:
    all_metrics.extend(r.get('metrics', []))

unique_metrics = list(set(all_metrics))
print(f"\n📈 Unique metrics calculated: {len(unique_metrics)}")
for metric in unique_metrics:
    print(f"   • {metric}")

high_confidence = len([r for r in results if r.get('confidence') == 'high'])
print(f"\n🎯 High confidence analyses: {high_confidence}/{len(results)}")

print("\n" + "="*80)
print("✅ QUANTITATIVE ANALYSIS COMPLETE!")
print("="*80)

print("\n💡 Your financial analysis includes:")
print("   ✓ Inventory Turnover calculations")
print("   ✓ Cash Conversion Cycle analysis")
print("   ✓ Profitability metrics and trends")
print("   ✓ Cost structure comparison")
print("   ✓ Multi-year trend analysis")
print("   ✓ Comprehensive ABC vs XYZ comparison")



✨ FINANCIAL ANALYSIS SUMMARY

📊 Total financial subtasks: 10
✅ Successful: 10
❌ Failed: 0

📈 Unique metrics calculated: 0

🎯 High confidence analyses: 10/10

✅ QUANTITATIVE ANALYSIS COMPLETE!

💡 Your financial analysis includes:
   ✓ Inventory Turnover calculations
   ✓ Cash Conversion Cycle analysis
   ✓ Profitability metrics and trends
   ✓ Cost structure comparison
   ✓ Multi-year trend analysis
   ✓ Comprehensive ABC vs XYZ comparison


# 📊 Financial Analysis for ABC vs XYZ

## Subtask Results

### 1. Evaluate Inventory Turnover
- **Formula:** Cost of Goods Sold / Average Inventory
- **Interpretation:** Higher turnover indicates efficient inventory management.
- **Analysis:**  
  - ABC Inventory Turnover Ratio ≈ 3.19  
  - XYZ Inventory Turnover Ratio ≈ 6.30  
  - **Conclusion:** XYZ sells inventory faster than ABC, indicating more efficient inventory management.

---

### 2. Analyze Cash Conversion Cycle (CCC)
- **Formula:** Days Inventory Outstanding + Days Sales Outstanding - Days Payable Outstanding
- **Interpretation:** Shorter cycle indicates better liquidity and efficiency.
- **Analysis:**  
  - ABC CCC ≈ 151 days  
  - XYZ CCC: N/A (some inventory/receivable/payable data missing)  
  - **Conclusion:** ABC has a longer cash conversion cycle than XYZ, suggesting slower working capital efficiency.

---

### 3. Assess Gross Margin
- **Formula:** (Revenue - COGS) / Revenue
- **Interpretation:** Higher margin indicates better pricing and cost control.
- **Analysis:**  
  - ABC Gross Margin ≈ 52.63%  
  - XYZ Gross Margin ≈ 44.64%  
  - **Conclusion:** ABC has higher profitability after COGS.

---

### 4. Evaluate Net Profit Margin
- **Formula:** Net Profit / Revenue
- **Interpretation:** Higher margin indicates better overall financial health.
- **Analysis:**  
  - ABC Net Profit Margin ≈ 6.9–7.7%  
  - XYZ Net Profit Margin ≈ 0%  
  - **Conclusion:** ABC is profitable; XYZ is not generating net profits.

---

### 5. Analyze SG&A %
- **Formula:** (SG&A Expenses / Revenue) × 100
- **Interpretation:** Lower percentage indicates better cost control.
- **Analysis:**  
  - ABC SG&A % ≈ 36.6–47.4%  
  - XYZ SG&A % ≈ 48.1–54.2%  
  - **Conclusion:** ABC has more efficient operating expenses relative to revenue.

---

### 6. Assess ROI
- **Formula:** (Net Profit / Total Investment) × 100
- **Analysis:**  
  - ABC ROI ≈ 100%  
  - XYZ ROI ≈ -8.86%  
  - **Conclusion:** ABC is generating positive returns; XYZ is experiencing losses.

---

### 7. Analyze Operating Margin
- **Formula:** Operating Income / Revenue
- **Analysis:**  
  - ABC Operating Margin ≈ 6.9–7.7%  
  - XYZ Operating Margin ≈ 0%  
  - **Conclusion:** ABC's core operations are profitable; XYZ is not.

---

### 8. Evaluate ROA (Return on Assets)
- **Formula:** Net Income / Total Assets
- **Analysis:**  
  - ROA could not be calculated due to missing net income data.  

---

### 9. Analyze ROE (Return on Equity)
- **Formula:** Net Income / Shareholders' Equity
- **Analysis:**  
  - ABC ROE ≈ 18.46%  
  - XYZ ROE ≈ -29.00%  
  - **Conclusion:** ABC provides positive returns to shareholders; XYZ is losing equity value.

---

### 10. Evaluate Revenue per Employee
- **Formula:** Revenue / Number of Employees
- **Analysis:**  
  - Cannot calculate due to missing employee data.  

---

## 📈 Comprehensive Comparison (Selected Metrics)

| Metric                     | ABC (2022)       | XYZ (2022)      | Notes |
|-----------------------------|----------------|----------------|-------|
| Gross Margin                | 50.7–52.63%    | 43.9–44.64%    | ABC more profitable after COGS |
| Net Profit Margin           | 6.9–7.7%       | 0%             | ABC profitable; XYZ not |
| Operating Margin            | 6.9–7.7%       | 0%             | ABC profitable operations |
| Inventory Turnover          | 3.19–3.25      | 6.3            | XYZ moves inventory faster |
| Asset Turnover              | 1.23–1.38      | 1.58–1.65      | XYZ slightly more efficient |
| CCC (days)                  | 151             | N/A            | ABC slower cash conversion |
| SG&A % of Revenue           | 36.6–47.4%     | 31.7–54.2%     | ABC better cost control |
| COGS % of Revenue           | 47.2–49.2%     | 54.9–56.1%     | XYZ higher cost of goods |
| ROI                         | 100%           | -8.86%         | ABC positive return; XYZ negative |
| ROE                         | 18.46%         | -29%           | ABC profitable for shareholders |

**Trends:**  
- ABC shows growth in revenue and net income, improving profitability margins.  
- XYZ shows declining or stagnant revenue, with negative profitability and ROE.  
- ABC has higher operating efficiency (gross/net/operating margins), while XYZ has faster inventory turnover but weaker overall financial health.
